File: kotlin/src/main/kotlin/dev/krom/parser/Daemon.kt
package dev.krom.parser

import dev.krom.parser.ipc.*
import dev.krom.parser.state.FileStateManager
import dev.krom.parser.treesitter.ParserEngine
import kotlinx.serialization.json.*

/**
 * Language-agnostic krom-parser daemon.
 * See /api.md for protocol specification.
 */
class Daemon(
    private val transport: Transport,
    private val engine: ParserEngine,
) {
    private val stateManager = FileStateManager(engine, engine.registry())
    private val json = Json { ignoreUnknownKeys = true }

    fun run() {
        transport.write(
            json.encodeToString(
                Response.serializer(),
                Response(id = 0, result = JsonPrimitive("krom-parser ready")),
            ),
        )

        while (true) {
            val raw = transport.read() ?: break
            val response = handleMessage(raw)
            if (response != null) {
                transport.write(json.encodeToString(Response.serializer(), response))
            }
        }
    }

    private fun handleMessage(raw: String): Response? {
        val request = try {
            json.decodeFromString(Request.serializer(), raw)
        } catch (e: Exception) {
            return Response(id = -1, error = ErrorInfo(-32700, "Parse error: ${e.message}"))
        }

        return try {
            val result = route(request.method, request.params)
            Response(id = request.id, result = result)
        } catch (e: Exception) {
            Response(id = request.id, error = ErrorInfo(-1, e.message ?: "Unknown error"))
        }
    }

    private fun route(method: String, params: JsonObject): JsonElement = when (method) {
        "ping" -> JsonPrimitive("pong")
        "shutdown" -> JsonPrimitive("ok").also {
            Thread { Thread.sleep(50); System.exit(0) }.start()
        }
        "listLanguages" -> json.encodeToJsonElement(
            kotlinx.serialization.builtins.ListSerializer(LanguageInfo.serializer()),
            engine.registry().listLanguages().map { m ->
                LanguageInfo(
                    id = m["id"] as String,
                    name = m["name"] as String,
                    version = m["version"] as String,
                    queries = m["queries"] as List<String>,
                    loaded = m["loaded"] as Boolean,
                )
            },
        )
        "parseFile" -> handleParseFile(params)
        "updateFile" -> handleUpdateFile(params)
        "closeFile" -> handleCloseFile(params)
        "getNodeAtPosition" -> handleGetNodeAtPosition(params)
        "getStructure" -> handleGetStructure(params)
        "getHighlights" -> handleGetHighlights(params)
        "runQuery" -> handleRunQuery(params)
        else -> throw IllegalArgumentException("Unknown method: $method")
    }

    private fun handleParseFile(params: JsonObject): JsonElement {
        val fileId = params.string("fileId")
        val content = params.string("content")
        val languageId = params.string("languageId")

        val state = stateManager.parseFile(fileId, content, languageId)
            ?: return JsonObject(
                mapOf(
                    "success" to JsonPrimitive(false),
                    "error" to JsonPrimitive("Unknown language: $languageId"),
                ),
            )

        return JsonObject(
            mapOf(
                "success" to JsonPrimitive(true),
                "version" to JsonPrimitive(state.version),
            ),
        )
    }

    private fun handleUpdateFile(params: JsonObject): JsonElement {
        val fileId = params.string("fileId")
        val content = params.string("content")

        val state = stateManager.updateFile(fileId, content)
            ?: return JsonObject(
                mapOf(
                    "success" to JsonPrimitive(false),
                    "error" to JsonPrimitive("File not open: $fileId"),
                ),
            )

        return JsonObject(
            mapOf(
                "success" to JsonPrimitive(true),
                "version" to JsonPrimitive(state.version),
            ),
        )
    }

    private fun handleCloseFile(params: JsonObject): JsonElement {
        stateManager.closeFile(params.string("fileId"))
        return JsonPrimitive("ok")
    }

    private fun handleGetNodeAtPosition(params: JsonObject): JsonElement {
        val node = stateManager.getNodeAtPosition(
            params.string("fileId"),
            params.int("line"),
            params.int("column"),
        ) ?: return JsonNull
        return json.encodeToJsonElement(NodeInfo.serializer(), node)
    }

    private fun handleGetStructure(params: JsonObject): JsonElement {
        val symbols = stateManager.getStructure(params.string("fileId"))
        return json.encodeToJsonElement(
            kotlinx.serialization.builtins.ListSerializer(SymbolInfo.serializer()),
            symbols,
        )
    }

    private fun handleGetHighlights(params: JsonObject): JsonElement {
        val result = stateManager.getHighlights(params.string("fileId"))
        return json.encodeToJsonElement(HighlightResult.serializer(), result)
    }

    private fun handleRunQuery(params: JsonObject): JsonElement {
        val nodes = stateManager.runQuery(params.string("fileId"), params.string("query"))
        return json.encodeToJsonElement(
            kotlinx.serialization.builtins.ListSerializer(NodeInfo.serializer()),
            nodes,
        )
    }

    private fun JsonObject.string(key: String): String =
        (this[key] as? JsonPrimitive)?.content
            ?: throw IllegalArgumentException("Missing param: $key")

    private fun JsonObject.int(key: String): Int =
        (this[key] as? JsonPrimitive)?.int
            ?: throw IllegalArgumentException("Missing param: $key")
}

File: go/internal/daemon/daemon.go
package daemon

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/daslaller/krom-parser/go/internal/ipc"
	"github.com/daslaller/krom-parser/go/internal/plugin"
	"github.com/daslaller/krom-parser/go/internal/state"
	"github.com/daslaller/krom-parser/go/internal/treesitter"
)

// Daemon is the language-agnostic krom-parser process. See /api.md.
type Daemon struct {
	transport *ipc.Transport
	engine    *treesitter.Engine
	plugins   []*plugin.Plugin
	state     *state.Manager
}

func New(transport *ipc.Transport, engine *treesitter.Engine, plugins []*plugin.Plugin) *Daemon {
	return &Daemon{
		transport: transport,
		engine:    engine,
		plugins:   plugins,
		state:     state.NewManager(engine, plugins),
	}
}

func (d *Daemon) Run() error {
	if err := ipc.WriteReady(d.transport); err != nil {
		return err
	}

	for {
		req, err := ipc.ReadRequest(d.transport)
		if err != nil {
			return nil // EOF — clean shutdown
		}
		resp := d.handle(req)
		if err := ipc.WriteResponse(d.transport, resp); err != nil {
			return err
		}
	}
}

func (d *Daemon) handle(req *ipc.Request) ipc.Response {
	result, err := d.route(req.Method, req.Params)
	if err != nil {
		return ipc.Response{ID: req.ID, Error: &ipc.ErrorInfo{Code: -1, Message: err.Error()}}
	}
	return ipc.Response{ID: req.ID, Result: result}
}

func (d *Daemon) route(method string, params map[string]interface{}) (interface{}, error) {
	switch method {
	case "ping":
		return "pong", nil
	case "shutdown":
		go func() {
			time.Sleep(50 * time.Millisecond)
			os.Exit(0)
		}()
		return "ok", nil
	case "listLanguages":
		return d.listLanguages(), nil
	case "parseFile":
		return d.parseFile(params)
	case "updateFile":
		return d.updateFile(params)
	case "closeFile":
		return d.closeFile(params)
	case "getNodeAtPosition":
		return d.getNodeAtPosition(params)
	case "getStructure":
		return d.getStructure(params)
	case "getHighlights":
		return d.getHighlights(params)
	case "runQuery":
		return d.runQuery(params)
	default:
		return nil, fmt.Errorf("unknown method: %s", method)
	}
}

func (d *Daemon) listLanguages() []map[string]interface{} {
	loaded := map[string]bool{}
	for _, p := range d.plugins {
		loaded[p.Manifest.ID] = d.engine.HasLanguage(p.Manifest.ID)
	}
	var out []map[string]interface{}
	for _, p := range d.plugins {
		out = append(out, p.ToLanguageInfo(loaded[p.Manifest.ID]))
	}
	return out
}

func (d *Daemon) parseFile(params map[string]interface{}) (map[string]interface{}, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	content, _ := ipc.ParamString(params, "content")
	languageID, _ := ipc.ParamString(params, "languageId")

	state, err := d.state.ParseFile(fileID, languageID, content)
	if err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   err.Error(),
		}, nil
	}
	return map[string]interface{}{
		"success": true,
		"version": state.Version,
	}, nil
}

func (d *Daemon) updateFile(params map[string]interface{}) (map[string]interface{}, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	content, _ := ipc.ParamString(params, "content")

	state, err := d.state.UpdateFile(fileID, content)
	if err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   err.Error(),
		}, nil
	}
	return map[string]interface{}{
		"success": true,
		"version": state.Version,
	}, nil
}

func (d *Daemon) closeFile(params map[string]interface{}) (string, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	d.state.CloseFile(fileID)
	return "ok", nil
}

func (d *Daemon) getNodeAtPosition(params map[string]interface{}) (interface{}, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	line, _ := ipc.ParamInt(params, "line")
	column, _ := ipc.ParamInt(params, "column")
	node := d.state.GetNodeAtPosition(fileID, line, column)
	if node == nil {
		return nil, nil
	}
	return node, nil
}

func (d *Daemon) getStructure(params map[string]interface{}) ([]ipc.SymbolInfo, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	return d.state.GetStructure(fileID), nil
}

func (d *Daemon) getHighlights(params map[string]interface{}) (ipc.HighlightResult, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	return d.state.GetHighlights(fileID), nil
}

func (d *Daemon) runQuery(params map[string]interface{}) ([]ipc.NodeInfo, error) {
	fileID, _ := ipc.ParamString(params, "fileId")
	query, _ := ipc.ParamString(params, "query")
	return d.state.RunQuery(fileID, query), nil
}

// HandleRaw is used in tests with raw JSON body.
func (d *Daemon) HandleRaw(body []byte) ipc.Response {
	var req ipc.Request
	if err := json.Unmarshal(body, &req); err != nil {
		return ipc.Response{ID: -1, Error: &ipc.ErrorInfo{Code: -32700, Message: "parse error"}}
	}
	if req.Params == nil {
		req.Params = map[string]interface{}{}
	}
	return d.handle(&req)
}
